File: problem solving/datastructures/graph-theory/lowest-common-ancestor.c

// https://www.hackerrank.com/challenges/binary-search-tree-lowest-common-ancestor/problem
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>

struct node {
    
    int data;
    struct node *left;
    struct node *right;
  
};

struct node* insert( struct node* root, int data ) {
		
	if(root == NULL) {
	
        struct node* node = (struct node*)malloc(sizeof(struct node));

        node->data = data;

        node->left = NULL;
        node->right = NULL;
        return node;
	  
	} else {
      
		struct node* cur;
		
		if(data <= root->data) {
            cur = insert(root->left, data);
            root->left = cur;
		} else {
            cur = insert(root->right, data);
            root->right = cur;
		}
	
		return root;
	}
}

/* you only have to complete the function given below.  
node is defined as  

struct node {
    
    int data;
    struct node *left;
    struct node *right;
  
};

*/
struct node *lca( struct node *root, int v1, int v2 ) {
        struct node *currentNode = root;
        while(currentNode != NULL) {
            if(v1 < currentNode->data && v2 < currentNode->data) {
                currentNode = currentNode->left;
            }
            else if (v1 > currentNode->data && v2 > currentNode->data) {
                currentNode = currentNode->right;
            }
            else {
                break;
            }
        }
        return currentNode;
}


int main() {
  
    struct node* root = NULL;
    
    int t;
    int data;

    scanf("%d", &t);

    while(t-- > 0) {
        scanf("%d", &data);
        root = insert(root, data);
    }
  	int v1;
  	int v2;
  
  	scanf("%d%d", &v1, &v2);
	struct node *ans = lca(root, v1, v2);
  	printf("%d", ans->data);
  	
    return 0;
}

File: problem solving/algorithms-approaches/dfs/lowestCommonAncestor.py
'''
    problem: https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/
    concepts: recursion, dfs, binary tree
    performance: 66.36% runtime, 53.22% memory
'''
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        def getLca(node):
            if not node:
                return None
            if node in [p, q]:
                return node
            l, r = getLca(node.left), getLca(node.right)
            if l and r:
                return node
            return l or r
        return getLca(root)